File: algorithms/math/segmented-sieve/java/SegmentedSieve.java
import java.util.ArrayList;
import java.util.List;

public class SegmentedSieve {
    public static int[] segmentedSieve(int low, int high) {
        if (high < 2 || high < low) {
            return new int[0];
        }
        low = Math.max(low, 2);

        int limit = (int) Math.sqrt(high);
        boolean[] prime = new boolean[limit + 1];
        java.util.Arrays.fill(prime, true);
        prime[0] = false;
        if (limit >= 1) {
            prime[1] = false;
        }

        List<Integer> basePrimes = new ArrayList<>();
        for (int p = 2; p <= limit; p++) {
            if (prime[p]) {
                basePrimes.add(p);
                for (int multiple = p * p; multiple <= limit; multiple += p) {
                    prime[multiple] = false;
                }
            }
        }

        boolean[] mark = new boolean[high - low + 1];
        java.util.Arrays.fill(mark, true);
        for (int p : basePrimes) {
            int start = Math.max(p * p, ((low + p - 1) / p) * p);
            for (int value = start; value <= high; value += p) {
                mark[value - low] = false;
            }
        }

        int count = 0;
        for (boolean isPrime : mark) {
            if (isPrime) {
                count++;
            }
        }

        int[] result = new int[count];
        int index = 0;
        for (int i = 0; i < mark.length; i++) {
            if (mark[i]) {
                result[index++] = low + i;
            }
        }
        return result;
    }
}

File: algorithms/math/segmented-sieve/swift/SegmentedSieve.swift
func segmentedSieve(_ low: Int, _ high: Int) -> [Int] {
    if high < 2 || low > high { return [] }

    let start = max(2, low)
    let limit = Int(Double(high).squareRoot())
    var isPrimeSmall = [Bool](repeating: true, count: max(2, limit + 1))
    var primes: [Int] = []

    if limit >= 2 {
        for value in 2...limit {
            if isPrimeSmall[value] {
                primes.append(value)
                var multiple = value * value
                while multiple <= limit {
                    isPrimeSmall[multiple] = false
                    multiple += value
                }
            }
        }
    }

    var isPrimeRange = [Bool](repeating: true, count: high - start + 1)
    for prime in primes {
        var multiple = max(prime * prime, ((start + prime - 1) / prime) * prime)
        while multiple <= high {
            isPrimeRange[multiple - start] = false
            multiple += prime
        }
    }

    var result: [Int] = []
    for value in start...high {
        if isPrimeRange[value - start] {
            result.append(value)
        }
    }
    return result
}
